flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/physical/stream/StreamExecFirstLastRowRule.scala

```diff
@@ -0,0 +1,207 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.rules.physical.stream
+
+import org.apache.flink.table.calcite.FlinkTypeFactory
+import org.apache.flink.table.plan.`trait`.FlinkRelDistribution
+import org.apache.flink.table.plan.nodes.FlinkConventions
+import org.apache.flink.table.plan.nodes.calcite.{ConstantRankRange, RankType}
+import org.apache.flink.table.plan.nodes.logical.{FlinkLogicalRank, FlinkLogicalSort}
+import org.apache.flink.table.plan.nodes.physical.stream.{StreamExecFirstLastRow, StreamExecRank, StreamExecSortLimit}
+import org.apache.flink.table.plan.util.FlinkRelOptUtil
+
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall}
+import org.apache.calcite.rel.`type`.RelDataType
+import org.apache.calcite.rel.convert.ConverterRule
+import org.apache.calcite.rel.core.Sort
+import org.apache.calcite.rel.{RelCollation, RelNode}
+
+/**
+  * Rule that matches [[FlinkLogicalSort]] which is sorted by proc-time attribute and
+  * fetches only one record started from 0, and converts it to [[StreamExecFirstLastRow]].
+  *
+  * NOTES: Queries that can be converted to [[StreamExecFirstLastRow]] could be converted to
+  * [[StreamExecSortLimit]] too. [[StreamExecFirstLastRow]] is more efficient than
+  * [[StreamExecSortLimit]] due to mini-batch and less state access.
+  *
+  * e.g.
+  * 1. ''SELECT a FROM MyTable ORDER BY proctime LIMIT 1'' will be converted to FirstRow
```

Comment:
If i understand `FirstRow` and `LastRow` correctly, they are not prepared for this case...

Reply:
yes, this case is rare. I think this case does not need to convert to `StreamExecRowDeduplicate` now.